core/unit_test/TestRangePolicyConstructors.hpp

```diff
@@ -0,0 +1,391 @@
+      ASSERT_EQ(p1.end(),        p2.end());
+      ASSERT_EQ(p1.chunk_size(), p2.chunk_size());
+    }
+  }
+
+  void test_worktag() {
+    struct WorkTag1 {};
+    struct WorkTag2 {};
+
+    using policy_t_no_worktag = Kokkos::RangePolicy<>;
+    using policy_t_worktag1 = Kokkos::Impl::WorkTagTrait::policy_with_trait<policy_t_no_worktag, WorkTag1>;
+    using policy_t_worktag2 = Kokkos::Impl::WorkTagTrait::policy_with_trait<policy_t_worktag1, WorkTag2>;
+
+    policy_t_no_worktag p0;
+    static_assert(std::is_void<typename decltype(p0)::work_tag>::value, "");
+
+    auto p1 = policy_t_worktag1(p0);
+    static_assert(std::is_same<typename decltype(p1)::work_tag, WorkTag1>::value, "");
+
+    auto p2 = policy_t_worktag2(p1);
+    static_assert(std::is_same<typename decltype(p2)::work_tag, WorkTag2>::value, "");
+
+    // NOTE line directly below does not currently compile
+    //using policy_t_void_tag = Kokkos::Impl::WorkTagTrait::policy_with_trait<policy_t_no_worktag, void>;
+    //auto p3 = policy_t_void_tag(p2);
+    //static_assert(std::is_void<decltype(p3)::work_tag>::value, "");
+  }
+
+  void test_prefer_desired_occupancy() {
+    using policy_t = Kokkos::RangePolicy<ExecutionSpace>;
+    policy_t policy;
+
+    static_assert(!policy_t::experimental_contains_desired_occupancy, "");
+
+    // MaximizeOccupancy -> MaximizeOccupancy
+    auto const policy_still_no_occ = Kokkos::Experimental::prefer(
+        policy, Kokkos::Experimental::MaximizeOccupancy{});
+    static_assert(!decltype(policy_still_no_occ)::experimental_contains_desired_occupancy, "");
+
+    // MaximizeOccupancy -> DesiredOccupancy
+    auto const policy_with_occ = Kokkos::Experimental::prefer(
+        policy_still_no_occ, Kokkos::Experimental::DesiredOccupancy{33});
+    static_assert(decltype(policy_with_occ)::experimental_contains_desired_occupancy, "");
+    EXPECT_EQ(policy_with_occ.impl_get_desired_occupancy().value(), 33);
+
+    // DesiredOccupancy -> DesiredOccupancy
+    auto const policy_change_occ = Kokkos::Experimental::prefer(
+        policy_with_occ, Kokkos::Experimental::DesiredOccupancy{24});
+    static_assert(decltype(policy_change_occ)::experimental_contains_desired_occupancy, "");
+    EXPECT_EQ(policy_change_occ.impl_get_desired_occupancy().value(), 24);
+
+    // DesiredOccupancy -> DesiredOccupancy w/ hint
+    auto policy_with_occ_and_hint = Kokkos::Experimental::require(
+        policy_change_occ, Kokkos::Experimental::WorkItemProperty::HintLightWeight);
+    EXPECT_EQ(policy_with_occ_and_hint.impl_get_desired_occupancy().value(), 24);
+
+    // DesiredOccupancy -> MaximizeOccupancy
+    auto const policy_drop_occ = Kokkos::Experimental::prefer(
+        policy_with_occ_and_hint, Kokkos::Experimental::MaximizeOccupancy{});
+    static_assert(!decltype(policy_drop_occ)::experimental_contains_desired_occupancy, "");
+  }
+};
+
+TEST(TEST_CATEGORY, range_policy_construction) {
+  TestRangePolicyConstructors<TEST_EXECSPACE>();
```

Comment:
Is it better to have the constructor called, which calls all the function? Or have different `TEST() { }` instances for each function in the struct?